tests/test_publish_dafsa.py

```diff
@@ -0,0 +1,148 @@
+                    self.assertIn("DAFSA Build Failed", str(cm.exception))
+
+
+class TestRemoteSettingsPublish(unittest.TestCase):
+    @responses.activate
+    def test_remote_settings_publish(self):
+        server = "https://kinto.dev.mozaws.net/v1"
+        record_uri = f"{server}/buckets/main-workspace/collections/public-suffix-list/records/tld-dafsa"  # noqa
+        attachment_uri = f"{record_uri}/attachment"
+
+        responses.add(
+            responses.PUT,
+            f"{server}/accounts/arpit73",
+            adding_headers={"Content-Type": "application/json"},
+            json='{"data": {"password": "pAsSwErD"}}',
+        )
+        client = Client(
+            server_url=server,
+            auth=("arpit73", "pAsSwErD"),
+            bucket=BUCKET_ID,
+            collection=COLLECTION_ID,
+        )
+        self.assertEqual(
+            client.get_endpoint("record", id=RECORD_ID),
+            "/buckets/dafsa-bucket/collections/public-suffix-list/records/tld-dafsa",
+        )
+
+        responses.add(
+            responses.POST,
+            attachment_uri,
+            adding_headers={"Content-Type": "multipart/form-data"},
+            files={"attachment": ("dafsa.bin", b"some binary data")},
+            json='{"commit_hash": "abc"}',
+        )
+        responses.add(
+            responses.PATCH,
+            record_uri,
+            adding_headers={"Content-Type": "application/json"},
+            json='{"data": {"status": "to-review"}}',
+        )
+
+        with tempfile.TemporaryDirectory() as tmp:
+            with open(f"{tmp}/dafsa.bin", "wb") as f:
+                f.write(b"some binary data")
+            remote_settings_publish(client, "abc", f"{tmp}/dafsa.bin")
+
+
+class TestPublishDafsa(unittest.TestCase):
+    event = {"server": "https://kinto.dev.mozaws.net/v1", "auth": "arpit73:pAsSwErD"}
+    record_uri = f"{{{event.get('server')}/buckets/main-workspace/collections/public-suffix-list/records/tld-dafsa}}"  # noqa
+
+    def setUp(self):
+        mocked = mock.patch("commands.publish_dafsa.prepare_dafsa")
+        self.addCleanup(mocked.stop)
+        self.mocked_prepare = mocked.start()
+
+        mocked = mock.patch("commands.publish_dafsa.remote_settings_publish")
+        self.addCleanup(mocked.stop)
+        self.mocked_publish = mocked.start()
+
+    @responses.activate
+    def test_prepare_and_publish_are_not_called_when_hash_matches(self):
+        responses.add(responses.GET, COMMIT_HASH_URL, json=[{"sha": "abc"}])
+        responses.add(
+            responses.GET, self.record_uri, json={"data": {"commit-hash": "abc"}}
+        )
+
+        publish_dafsa(self.event, context=None)
+
+        self.assertFalse(self.mocked_prepare.called)
+        self.assertFalse(self.mocked_publish.called)
+
+    @responses.activate
+    def test_KintoException_raised_when_fetching_failed(self):
+        responses.add(responses.GET, COMMIT_HASH_URL + "c", json=[{"sha": "abc"}])
```

Comment:
Why do you add `+ "c"` ? 

The whole point of the `responses` library is to control what response is obtained in the test.
You can thus use it to return a 404 or whatever!

```py
responses.add(responses.GET, COMMIT_HASH_URL, json={'error': 'not found'}, status=404) 
```